File: transformer/py/transformerlib/rng.py
# -*- coding: utf-8 -*-
"""난수 — splitmix64 로 씨앗을 펴고 xoshiro256** 로 뽑는다 (SPEC.md §1).

파이썬의 random 모듈을 쓰지 않는 까닭: C 가 같은 수열을 내야 한다.
random 의 메르센 트위스터는 C 로 옮길 수는 있지만, 씨앗을 펴는 방식과
정규분포를 뽑는 방식(가우스 캐시)까지 맞추기가 번거롭다. 짧고 전부
적어 둘 수 있는 생성기를 골랐다. 한 번 뽑기는 O(1) 시간·O(1) 공간.

파이썬 정수는 넘치지 않으므로 64비트 넘침을 & MASK 로 흉내 낸다.
"""
import math

MASK = (1 << 64) - 1
INV53 = 1.0 / 9007199254740992.0          # 2⁻⁵³


def _rotl(x, k):
    return ((x << k) | (x >> (64 - k))) & MASK


def splitmix64(x):
    """(다음 상태, 출력). 씨앗 하나를 64비트 넷으로 펴는 데 쓴다."""
    x = (x + 0x9E3779B97F4A7C15) & MASK
    z = x
    z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK
    z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK
    return x, z ^ (z >> 31)


class Rng(object):
    """xoshiro256** 생성기 하나. 쓰임마다 따로 만든다(SPEC §1.5)."""

    def __init__(self, seed):
        x = seed & MASK
        self.s = []
        for _ in range(4):
            x, z = splitmix64(x)
            self.s.append(z)

    def next(self):
        """64비트 정수 하나 (SPEC §1.2)."""
        s = self.s
        result = (_rotl((s[1] * 5) & MASK, 7) * 9) & MASK
        t = (s[1] << 17) & MASK
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = _rotl(s[3], 45)
        return result

    def uniform(self):
        """[0, 1) — 위 53비트만 쓴다. double 이 정확히 담는 폭이다."""
        return (self.next() >> 11) * INV53

    def normal(self):
        """표준정규 하나 — Box–Muller 의 코사인 쪽만.

        사인 쪽을 남겨 다음에 주면 한 번에 둘을 얻지만, 그 캐시가
        C 와 "몇 번째 호출인가" 를 맞춰야 하는 짐이 된다. 버린다.
        """
        u1 = 1.0 - self.uniform()          # (0, 1] — ln 0 을 피한다
        u2 = self.uniform()
        return math.sqrt(-2.0 * math.log(u1)) * math.cos(2.0 * math.pi
                                                          * u2)

    def randint(self, n):
        """0 ≤ r < n. ⌊uniform·n⌋ — 치우침은 2⁻⁵³·n 수준이라 무시."""
        if n <= 0:
            raise ValueError('randint(%d): n 은 양수여야 한다' % n)
        return int(self.uniform() * n)

    def permutation(self, n):
        """피셔–예이츠. O(n) 시간·O(n) 공간."""
        p = list(range(n))
        for i in range(n - 1, 0, -1):
            j = self.randint(i + 1)
            p[i], p[j] = p[j], p[i]
        return p

File: transformer/c/rng.c
/* rng.c — splitmix64 로 씨앗을 펴고 xoshiro256** 로 뽑는다 (SPEC §1).
 *
 * C 의 uint64_t 는 2⁶⁴ 에서 넘치는 것이 규격이다. 파이썬이 & MASK 로
 * 흉내 내던 일을 여기서는 언어가 해 준다. 한 번 뽑기 O(1).
 */
#include <math.h>

#include "rng.h"

/* -std=c99 에는 M_PI 가 없다. 파이썬 math.pi 의 repr 과 같은 double. */
#define TFS_PI 3.141592653589793

static uint64_t rotl(uint64_t x, int k)
{
    return (x << k) | (x >> (64 - k));
}

void tfs_rng_seed(tfs_rng *r, uint64_t seed)
{
    uint64_t x = seed;
    int i;

    for (i = 0; i < 4; i++) {
        uint64_t z = (x += 0x9E3779B97F4A7C15ULL);
        z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
        z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
        r->s[i] = z ^ (z >> 31);
    }
}

uint64_t tfs_rng_next(tfs_rng *r)
{
    uint64_t *s = r->s;
    uint64_t result = rotl(s[1] * 5, 7) * 9;
    uint64_t t = s[1] << 17;

    s[2] ^= s[0];
    s[3] ^= s[1];
    s[1] ^= s[2];
    s[0] ^= s[3];
    s[2] ^= t;
    s[3] = rotl(s[3], 45);
    return result;
}

/* [0, 1) — 위 53비트. 2⁻⁵³ 은 double 로 정확히 표현된다. */
double tfs_rng_uniform(tfs_rng *r)
{
    return (double)(tfs_rng_next(r) >> 11)
           * (1.0 / 9007199254740992.0);
}

/* Box–Muller 의 코사인 쪽만, 캐시 없이. u1 = 1 − uniform 은 (0, 1]. */
double tfs_rng_normal(tfs_rng *r)
{
    double u1 = 1.0 - tfs_rng_uniform(r);
    double u2 = tfs_rng_uniform(r);
    return sqrt(-2.0 * log(u1)) * cos(2.0 * TFS_PI * u2);
}

int tfs_rng_randint(tfs_rng *r, int n)
{
    return (int)(tfs_rng_uniform(r) * n);
}

/* 피셔–예이츠, i = n−1 에서 1 까지. O(n). */
void tfs_rng_permutation(tfs_rng *r, int *p, int n)
{
    int i;

    for (i = 0; i < n; i++)
        p[i] = i;
    for (i = n - 1; i > 0; i--) {
        int j = tfs_rng_randint(r, i + 1);
        int t = p[i];
        p[i] = p[j];
        p[j] = t;
    }
}
